### ISSUE TITLE
Speed up twisted.python.context by copying less data

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4960]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4960|
|Type|enhancement|
|Created|2011-03-15 20:48:36Z|
|Branch|https://github.com/twisted/twisted/tree/faster-callwithcontext-4960|

`ContextTracker.callWithContext` makes a copy of the current context and update it with the new context data.  This makes `callWithContext` somewhat expensive, even if no code actually loads anything from the context.

Contrariwise, it makes it easy to retrieve keys from the context, which most code doesn't bother to do.

We should reverse these, making `callWithContext` as cheap as we can, even at the cost of slowing down `getContext` somewhat.

<details><summary>Searchable metadata</summary>

```
trac-id__4960 4960
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_faster_callwithcontext_4960 branches/faster-callwithcontext-4960
branch_author__exarkun__fijal exarkun, fijal
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1300222116000000 1300222116000000
changetime__1300300269000000 1300300269000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [31152]) Branching to 'faster-callwithcontext-4960'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [31153]) Make the context stack into a chain that you have to follow, instead of a tower with everything piled on top

refs [#4960](https://github.com/twisted/twisted/issues/4960)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

This doesn't seem to change CPython performance measurably, but it speeds up some benchmarks on PyPy by... maybe... 5% or so (eyeball).

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli set owner to @cyli<br>@cyli set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli set owner to @exarkun<br>@cyli set status to `new`|
|-|-|

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/faster-callwithcontext-4960) are good.

The only concern I have is about the performance cost of exceptions, but this concern may not be valid if getContext most often gets the last context appended to the chain.  I'm not sure how getContext is usually used.

Otherwise, looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [31183]) Merge faster-callwithcontext-4960

Author: exarkun, fijal
Reviewer: cyli
Fixes: [#4960](https://github.com/twisted/twisted/issues/4960)

Speed up event dispatch (mostly on PyPy) by reducing the amount of copying
involved in the context/logging setup done for each event, at the expense
of making key lookup in the context potentially less efficient.
